### PR TITLE
Add code representation for interactions

### PR DESCRIPTION
**Pull request by wKich**
## Motivation

The second step of providing code representation for interactions to be able to show them in [interactive stories](https://storybook.js.org/blog/interactive-stories-beta/).

## Approach

Pass an action event into an action wrapper

## Issues

- Interactor's name might have space characters or any other restricted symbols so code representation will look like this
`text field("Password", { "enabled": true }).value()`
- The same problem with action/filter names
- Action arguments simply stringifying through `JSON.stringify`
- Don't stringify functions that passed to `assert` and `perform` methods

### CODE REVIEW

File: packages/core/src/serialize.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  ...restOptions
+}: ActionOptions): SerializedActionOptions {
+  let interactor = serializeInteractorOptions(options);
+  let ancestors = options.ancestors.map((ancestor) => serializeInteractorOptions(ancestor));
+  let args = "";
+  if ("filters" in restOptions) {
+    let filters: string[] = [];
+    for (let name in restOptions.filters) {
+      filters.push(`"${name}": ${matcherCode(restOptions.filters[name])}`);
+    }
```
Comment (jnicklas) [inline packages/core/src/serialize.ts:37]: Why not use `.map` here?

**Comment by jnicklas** [comment]
If I'm understanding this correctly, then we are generating the code description "up-front" at the moment we are creating the interaction. I think it might be nicer to defer this until we actually need it?

**Comment by cowboyd** [comment]
Should we do the lazy serialization in the `interaction.toString()` method?

File: packages/core/src/serialize.ts (hunk 1 above)
Comment (wKich) [reply packages/core/src/serialize.ts:37]: filters is an object, or do you mean `Object.entries().map`?

**Comment by wKich** [comment]
Yeah, that's good point

File: packages/globals/src/globals.ts [hunk 2]
```diff
@@ -1,19 +1,43 @@
-import { KeyboardLayout } from './keyboard-layout';
-
 /* eslint-disable @typescript-eslint/no-explicit-any */
-type ActionType = "interaction" | "check";
+import { KeyboardLayout } from "./keyboard-layout";
+
+export type ActionType = "interaction" | "check";
```
Comment (cowboyd) [inline packages/globals/src/globals.ts:4]: For consistency, this should probably be

```ts
export type InteractionType = "action" | "check"
```

**Review by cowboyd — COMMENTED** [review]
I'm kind of feeling like we should revise our text representations to be `"TextField"` instead of `"text field"` and make that the recommended format. What do you think?

File: packages/globals/src/globals.ts (hunk 2 above)
Comment (wKich) [reply packages/globals/src/globals.ts:4]: It was changed here https://github.com/thefrontside/interactors/pull/166

**Comment by wKich** [comment]
@cowboyd I like your suggestion about naming convention

File: packages/globals/src/globals.ts (hunk 2 above)
Comment (cowboyd) [reply packages/globals/src/globals.ts:4]: Maybe we can do this as a separate PR?

**Comment by cowboyd** [comment]
Could you go a little more into what do these issues represent? 

* Action arguments simply stringifying through JSON.stringify
* Don't stringify functions that passed to assert and perform methods

**Comment by wKich** [comment]
- Action arguments simply stringifying through JSON.stringify:

I mean, that I just did `JSON.stringify` on action arguments and didn't check type, so an argument might be any type, like `function`, `HTMLElement`, `RegExp` or whatever. And I chose don't add additional overwhelming transformation here.


- Don't stringify functions that passed to assert and perform methods:

And here, user is able to pass function as an argument in `assert` and `perform` methods, so I decided just ignore that function and don't stringify it, because we can't get source code of original function and if try to call `toString` we will get transpailed representation

File: packages/core/src/constructor.ts [hunk 3]
```diff
@@ -151,15 +152,21 @@ export function instantiateInteractor<E extends Element, F extends Filters<E>, A
...
-      return check(`${description(options)} asserts`, () => {
-        return converge(() => {
-          fn(resolver(options));
-        });
-      });
+      return check(
+        interaction(
+          `${description(options)} asserts`,
+          () => converge(() => { fn(resolver(options)) }),
+          serializeActionOptions({ type: "check", actionName: "assert", options }),
```
Comment (cowboyd) [inline packages/core/src/constructor.ts:167]: Can we just add these options directly to the `interaction` method and have it call `serializeActionOptions` inside itself?

File: packages/globals/src/globals.ts (hunk 2 above)
Comment (wKich) [reply packages/globals/src/globals.ts:4]: In that case it's easier to do it in this PR

**Review by cowboyd — COMMENTED** [review]
If every time that we need to create an interaction, we also need to pass it a result of `serializeInteractionOptions`, then doesn't it make sense to just pass the options directly to `interaction()` and have it do the serialization itself?

File: packages/core/src/constructor.ts (hunk 3 above)
Comment (wKich) [reply packages/core/src/constructor.ts]: Done

**Review by cowboyd — COMMENTED** [review]
Is there any reason not to put `.code` directly on the interaction instead of embedded in `.options.code` ?

**Comment by wKich** [comment]
No. There is no reason.

**Review by cowboyd — APPROVED** [review]
🎉 🥳 🥳 
Awesome interactions, and debuggable interactions just got a big leap forward!

**Pull request merged**